openmeter/entitlement/adapter/entitlement.go

```diff
@@ -166,6 +172,39 @@ func (a *entitlementDBAdapter) DeleteEntitlement(ctx context.Context, entitlemen
 	return err
 }
 
+func (a *entitlementDBAdapter) DeactivateEntitlement(ctx context.Context, entitlementID models.NamespacedID, at time.Time) error {
+	_, err := entutils.TransactingRepo[interface{}, *entitlementDBAdapter](
+		ctx,
+		a,
+		func(ctx context.Context, repo *entitlementDBAdapter) (interface{}, error) {
+			ent, err := repo.GetEntitlement(ctx, entitlementID)
+			if err != nil {
+				return nil, err
+			}
+
+			if ent.ActiveTo != nil {
+				return nil, fmt.Errorf("entitlement %s is already deactivated", entitlementID.ID)
+			}
+
+			c, err := repo.db.Entitlement.Update().
+				Where(db_entitlement.ID(entitlementID.ID), db_entitlement.Namespace(entitlementID.Namespace)).
+				SetActiveTo(at).
+				Save(ctx)
+			if err != nil {
+				return nil, err
+			}
+
+			if c != 1 {
+				return nil, fmt.Errorf("updated %d entitlements, which is not 1", c)
```

Comment:
Maybe return a not found error for c == 0

Reply:
line 182 already returns NotFound if not found
```go
ent, err := repo.GetEntitlement(ctx, entitlementID)
if err != nil {
  return nil, err
}
```